src/platform_impl/linux/x11/window.rs

```diff
@@ -809,6 +809,20 @@ impl UnownedWindow {
                 leap!(result).ignore_error();
             }
 
+            // Select XInput2 events
+            let mask = xinput::XIEventMask::MOTION
+                | xinput::XIEventMask::BUTTON_PRESS
+                | xinput::XIEventMask::BUTTON_RELEASE
+                | xinput::XIEventMask::ENTER
+                | xinput::XIEventMask::LEAVE
+                | xinput::XIEventMask::FOCUS_IN
+                | xinput::XIEventMask::FOCUS_OUT
+                | xinput::XIEventMask::TOUCH_BEGIN
+                | xinput::XIEventMask::TOUCH_UPDATE
+                | xinput::XIEventMask::TOUCH_END;
+            leap!(xconn.select_xinput_events(window.xwindow, super::ALL_MASTER_DEVICES, mask))
+                .ignore_error();
```

Comment:
Precisely the idea is to register the XInput events before mapping the window, because if you do it afterwards, you may miss XInput messages generated during the mapping, such as the FocusIn from the related bug.

Moving it to before the mapping guarantees that we don't miss those messages.

Reply:
@kchibisov Any feedback about this question?

Note that this is actually the change that fixes #2841, the most important change.

The second commit is just an improvement related to the same symptoms but an unrelated cause, that could be omitted if you feel unsure about it.

Reply:
The idea is to have the XInput2 events registered before the window is mapped. Else you risk losing the events generated during the mapping, such as the very first "set-focus".